src/Storages/MergeTree/GinIndexStore.cpp

```diff
@@ -0,0 +1,435 @@
+
+        return postings_list;
+    }
+}
+
+bool GinIndexStore::exists() const
+{
+    String id_file_name = getName() + GIN_SEGMENT_ID_FILE_TYPE;
+    return storage->exists(id_file_name);
+}
+
+UInt32 GinIndexStore::getNextIDRange(const String& file_name, size_t n)
+{
+    std::lock_guard<std::mutex> guard{gin_index_store_mutex};
+
+    if (!storage->exists(file_name))
+    {
+        std::unique_ptr<DB::WriteBufferFromFileBase> ostr = this->data_part_storage_builder->writeFile(file_name, DBMS_DEFAULT_BUFFER_SIZE, {});
+
+        const auto& int_type = DB::DataTypePtr(std::make_shared<DB::DataTypeUInt32>());
+        auto size_serialization = int_type->getDefaultSerialization();
+        size_serialization->serializeBinary(1, *ostr);
+        ostr->sync();
+    }
+
+    /// read id in file
+    UInt32 result = 0;
+    {
+        std::unique_ptr<DB::ReadBufferFromFileBase> istr = this->storage->readFile(file_name, {}, std::nullopt, std::nullopt);
+
+        Field field_rows;
+        const auto& size_type = DB::DataTypePtr(std::make_shared<DB::DataTypeUInt32>());
+        auto size_serialization = size_type->getDefaultSerialization();
+
+        size_type->getDefaultSerialization()->deserializeBinary(field_rows, *istr);
+        result = field_rows.get<UInt32>();
+    }
+    //save result+n
+    {
+        std::unique_ptr<DB::WriteBufferFromFileBase> ostr = this->data_part_storage_builder->writeFile(file_name, DBMS_DEFAULT_BUFFER_SIZE, {});
+
+        const auto& int_type = DB::DataTypePtr(std::make_shared<DB::DataTypeUInt32>());
+        auto size_serialization = int_type->getDefaultSerialization();
+        size_serialization->serializeBinary(result + n, *ostr);
+        ostr->sync();
+    }
+    return result;
+}
+
+UInt32 GinIndexStore::getNextRowIDRange(size_t n)
+{
+    UInt32 result =current_segment.next_row_id;
+    current_segment.next_row_id += n;
+    return result;
+}
+
+UInt32 GinIndexStore::getNextSegmentID()
+{
+    String sid_file_name = getName() + GIN_SEGMENT_ID_FILE_TYPE;
+    return getNextIDRange(sid_file_name, 1);
+}
+
+UInt32 GinIndexStore::getSegmentNum()
+{
+    if (cached_segment_num)
+        return cached_segment_num;
+
+    String sid_file_name = getName() + GIN_SEGMENT_ID_FILE_TYPE;
+    if (!storage->exists(sid_file_name))
+        return 0;
+    Int32 result = 0;
+    {
+        std::unique_ptr<DB::ReadBufferFromFileBase> istr = this->storage->readFile(sid_file_name, {}, std::nullopt, std::nullopt);
+
+        Field field_rows;
+        const auto& size_type = DB::DataTypePtr(std::make_shared<DB::DataTypeUInt32>());
+        auto size_serialization = size_type->getDefaultSerialization();
+
+        size_type->getDefaultSerialization()->deserializeBinary(field_rows, *istr);
+        result = field_rows.get<UInt32>();
+    }
+
+    cached_segment_num = result - 1;
+    return cached_segment_num;
+}
+
+bool GinIndexStore::needToWrite() const
+{
+    assert(max_digestion_size > 0);
```

Comment:
Not clear why outside callers are supposed to decide when GinIndexStore should write. Why can GinIndexStore not decide that by itself?

Reply:
There are two constructors for GinIndexStore, one is for creating index and another one is for reading index(you can see it by searching `std::make_shared<GinIndexStore>`). So the assert is to make sure GinIndexStore was constructed by the constructor for creating index.
Also the inverted index is a skipping index which has to go through the digesting process of generic skipping index and also it needs to do segmentation when required, so that's why it cannot decide when it writes.